### ISSUE TITLE
Creation of dataset with spaces in its name : mounted with false attribute under Debian Buster 4.19.0-12-amd64 x86_64

### ISSUE THREAD

**Issue by Bernard-Limogne** [openzfs/zfs#11182]
Under Debian Buster (uname -a = "Linux antec 4.19.0-12-amd64 #1 SMP Debian 4.19.152-1 (2020-10-18) x86_64 GNU/Linux"), creation of a dataset ("zfs create -o compression=lz4 -o normalization=formD -o utf8only=on -o dnodesize=auto -o relatime=on -o acltype=posixacl 'tank/name with blanks'").
mount gives : "tank/name on /mnt/tank/name with blanks type zfs (rw,relatime,with)"
The dataset cannot be mounted by : "zfs mount 'tank/name with blanks'" ==> "cannot mount 'tank/name with blanks': mountpoint or dataset busy" (normal diagnostic)

The same test on FreeBSD13 works (with same ZFS).

### System information

Type | Version/Name
 --- | ---
Distribution Name	| 
Distribution Version	|
Linux Kernel	|4.19.0-12
Architecture	|amd64 x86_64
ZFS Version	|OpenZFS 2.0.0-rc5
SPL Version	| none

**Comment by misterbigstuff** [comment]
this has never worked on Linux, as was discovered when support for fs with spaces in the name were introduced to Dracut initrd code.

**Comment by Bernard-Limogne** [comment]
I have forgotten 2 things.
The dataset cannot be mounted or unmounted via ZFS.
It cannot be destroyed before unmounted by the system command umount.

**Comment by Bernard-Limogne** [comment]
@misterbigstuff 
The same test works on the same Debian with zfs-0.8.2-1.

**Comment by behlendorf** [comment]
@Bernard-Limogne thanks for catching this regression, it appears to have been an accidental side effect of the #10839 PR..   I've opened #11187 to address it.  If you're comfortable applying the patch I'd be great if you could verify this resolves the issue on Debian.

**Comment by Bernard-Limogne** [comment]
> @Bernard-Limogne thanks for catching this regression, it appears to have been an accidental side effect of the #10839 PR.. I've opened #11187 to address it. If you're comfortable applying the patch I'd be great if you could verify this resolves the issue on Debian.

**Comment by Bernard-Limogne** [comment]
@behlendorf 
Hi Brian
I've tested your patch (in module/os/linux/zfs/zpl_super.c lines [193,204]).
Same result in my context :
 - last zfs-2.0.0-rc5.tar.gz, patched, compiled, linked and installed
 - Debian Buster (uname -a : "Linux buster 4.19.0-12-amd64 #1 SMP Debian 4.19.152-1 (2020-10-18) x86_64 GNU/Linux").

I have 2 questions.
1) when you write "seq_puts(seq, "\\040");" at line 200, you insert, in place of a space, a C string (so with an ending binary 0) : is it normal ? (It's been over 20 years since I wrote a line of C code... so my question is perhaps stupid)
2) i don't understand what seq_puts() is for; but  shouldn't line 193 be overwritten by the patch?
Regards

**Comment by behlendorf** [comment]
> Same result in my context 

By this you mean it resolved the issue, correct?  Since you closed the issue that seems like it's the case.

> when you write "seq_puts(seq, "\040");" at line 200, you insert, in place of a space, a C
> but shouldn't line 193 be overwritten by the patch

No problem.  That's right.  `seq_puts()` in a kernel helper function which will output the passed string up to, but not including, the terminating NULL (0) character.  It then advances the pointer in the output `seq_file` so we're ready to write the next chunk.  All we're doing here is iterating over the dataset name and replacing spaces with `"\040"` (the octal escape code for a space) and leaving all other characters the same.

This might seem a little strange but this function is what is used to generate the device name in the "/etc/mtab" file.  These days that file is really just a symlink to `/proc/self/mounts`.  The `zfs` command uses the standard [getmntent(3)](https://man7.org/linux/man-pages/man3/getmntent.3.html) library function to parse the file to determine what's mounted and where.  Since white space was used as a separator in this file the library function expects spaces to be encoded as octal.  Here's the relevant chunk from the man page:

```
       The mntent structure is defined in <mntent.h> as follows:

           struct mntent {
               char *mnt_fsname;   /* name of mounted filesystem */
               char *mnt_dir;      /* filesystem path prefix */
               char *mnt_type;     /* mount type (see mntent.h) */
               char *mnt_opts;     /* mount options (see mntent.h) */
               int   mnt_freq;     /* dump frequency in days */
               int   mnt_passno;   /* pass number on parallel fsck */
           };

       Since fields in the mtab and fstab files are separated by whitespace,
       octal escapes are used to represent the characters space (\040), tab
       (\011), newline (\012), and backslash (\\) in those files when they
       occur in one of the four strings in a mntent structure.  The routines
       addmntent() and getmntent() will convert from string representation
       to escaped representation and back.  When converting from escaped
       representation, the sequence \134 is also converted to a backslash.
```

Since dataset names can't contain tabs, newlines, or backslashes we only need to handle translating the space.  You can run `cat /proc/self/mounts` to verify the string was generated correctly.  Prior to us adding our own custom function for this the default kernel function took care of escaping the spaces which why this works in earlier releases.  Thanks again for catching this.

**Comment by Bernard-Limogne** [comment]
@behlendorf Hi Brian
Thanks for these explanations. But the patch doesn't solve the problem that may be elsewhere.
There are 2 anomalies.
1) The file /proc/self/mounts says this : 'tank/name with blanks /mnt/tank/name\040with\040blanks zfs rw,relatime,xattr,posixacl 0 0' Thus the dataset seems to be correctly defined.
But the command mount says : 'tank/name on /mnt/tank/name with blanks type zfs (rw,relatime,with)'
The first space has cut the name and the word 'with' was considered an attribute.
2) For the system (see the command mount), the dataset is mounted; for ZFS, it isn't... and more : it can't be "ZFS mounted" (error 'mountpoint or dataset is busy').
So the path is insufficient.
And this issue must remain OPEN
Regards

**Comment by behlendorf** [comment]
@Bernard-Limogne with the patch applied I'm unable to reproduce the issue on Debian Buster using your test case.  Have I overlooked something?  I wasn't able to reproduce either of the 2 anomalies.  See my test output below:

```bash
# Create the pool and filesystem
$ truncate -s 1G /var/tmp/vdev
$ zpool create tank /var/tmp/vdev
$ zfs create -o compression=lz4 -o normalization=formD -o utf8only=on -o dnodesize=auto -o relatime=on -o acltype=posixacl "tank/name with blanks"

# Check the output of `mount(8)` and `zfs mount`
$ mount -t zfs
tank on /tank type zfs (rw,xattr,noacl)
tank/name with blanks on /tank/name with blanks type zfs (rw,relatime,xattr,posixacl)

$ zfs mount
tank                            /tank
tank/name with blanks           /tank/name with blanks

# Check `/proc/self/mounts` is properly escaped
$ cat /proc/self/mounts | grep zfs
tank /tank zfs rw,xattr,noacl 0 0
tank/name\040with\040blanks /tank/name\040with\040blanks zfs rw,relatime,xattr,posixacl 0 0

# Expected error when the filesystem is already mounted
$ zfs mount 'tank/name with blanks'
cannot mount 'tank/name with blanks': filesystem already mounted

# Clean unmount and expected unmounted error.
$ zfs unmount 'tank/name with blanks'
$ zfs unmount 'tank/name with blanks'
cannot unmount 'tank/name with blanks': not currently mounted

# Clean mount
$ zfs mount 'tank/name with blanks'
$ zfs mount
tank                            /tank
tank/name with blanks           /tank/name with blanks
```

**Comment by Bernard-Limogne** [comment]
@behlendorf Hi Brian
Here is the result of your test in my context.
# Create the pool and filesystem
: truncate -s 1G /var/tmp/vdev
: zpool create tank /var/tmp/vdev
: zfs create -o compression=lz4 -o normalization=formD -o utf8only=on -o dnodesize=auto -o relatime=on -o acltype=posixacl "tank/name with blanks"

: mount -t zfs
tank/name on /tank/name with blanks type zfs (rw,relatime,with)

: cat /proc/self/mounts | grep zfs
tank/name with blanks /tank/name\040with\040blanks zfs rw,relatime,xattr,posixacl 0 0

:zfs mount 'tank/name with blanks'
cannot mount 'tank/name with blanks': mountpoint or dataset is busy

:zfs unmount 'tank/name with blanks'
cannot unmount 'tank/name with blanks': not currently mounted

:zfs mount 'tank/name with blanks'
cannot mount 'tank/name with blanks': mountpoint or dataset is busy

: zfs mount
tank      /tank

My result and yours are contradictory : let's see where the differences between our 2 contexts can be.
I omit the fact that the creator of the pool and the dataset is a normal user for you, while for me he is an administrator.
At me, Debian runs in a VM (VirtualBox) inside a Debian host (different VMs can run with differents OSes).
I installed Debian without ZFS and took the ZFS sources at https://github.com/openzfs/zfs/releases/tag/zfs-2.0.0-rc5 on the line ' zfs-2.0.0-rc5.tar.gz'.
I made the same modification as you in the source module/os/linux/zfs/zpl_super.c in which i did the same modification as you.
So what to think?
Regards

**Comment by behlendorf** [comment]
I performed my testing in a clean Buster image using the Debian provided [EC2 AMI](https://wiki.debian.org/Cloud/AmazonEC2Image/Buster).  To double check I ran through the test again today using the zfs-2.0.0-rc6.tar.gz from yesterday which includes this fix.  Again I didn't encounter any problems.

Are you by chance using musl libc in your Debian install?  Based on [this mailing list thread](https://inbox.vuxu.org/musl/HE1P192MB0010D60AAC4BA0F08F1A5644FB5E0@HE1P192MB0010.EURP192.PROD.OUTLOOK.COM/T/) it sounds like musl's implementation of `getmntent(3)` doesn't yet support unescaping special characters.  This would explain the symptoms your seeing.
